File: SharpOffice.Common.Tests/Configuration/ConfigurationFormatTest.cs
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Moq;
using NUnit.Framework;
using SharpOffice.Common.Configuration;
using SharpOffice.Core.Configuration;
using SharpOffice.Core.Formats;

namespace SharpOffice.Common.Tests.Configuration
{
    [TestFixture]
    public class ConfigurationFormatTest
    {
        public IConfiguration GetMockConfiguration(IEnumerable<KeyValuePair<string, object>> testData)
        {
            var mock = new Mock<IConfiguration>();
            var data = testData as IList<KeyValuePair<string, object>> ?? testData.ToList();
            mock.Setup(cfg => cfg.GetAllProperties()).Returns(data);
            return mock.Object;
        }

        public void ReadWriteInMemoryStreamTest<TFormat, TConfig>(IConfiguration configuration)
            where TFormat : IConfigurationFormat, new()
            where TConfig : IConfiguration, new()
        {
            var stream = new MemoryStream();
            var configFormat = new TFormat();
            configFormat.WriteConfiguration(configuration, stream);

            stream.Position = 0;
            var config = configFormat.ReadConfiguration<TConfig>(stream);
            foreach (var pair in configuration.GetAllProperties())
                if (!(pair.Value is IEnumerable) || pair.Value is string)
                    Assert.AreEqual(pair.Value, config.GetProperty<object>(pair.Key));
                else
                    foreach (var obj in config.GetProperty<IEnumerable>(pair.Key))
                        Assert.IsTrue(Has((IEnumerable) pair.Value, obj));
        }

        [Test]
        public void BinaryConfigurationFormatTest_WithMock()
        {
            ReadWriteInMemoryStreamTest<BinaryConfigurationFormat, TestConfiguration>(
                GetMockConfiguration(
                    new Dictionary<string, object> { { "int", 16 }, { "float", 3.2f } }));
        }
        
        [Test]
        public void BinaryConfigurationFormatTest_WithPropertyBasedConfiguration()
        {
            ReadWriteInMemoryStreamTest<BinaryConfigurationFormat, TestPropertyBasedConfiguration>(
                new TestPropertyBasedConfiguration()
                {
                    Integer = 12,
                    Text = "Warszawa",
                    Bits = new[] {true, false}
                });
        }

        [Test]
        public void YamlConfigurationFormatTest()
        {
            var config = new TestPropertyBasedConfiguration();
            config.SetProperty("Integer", 16);
            config.SetProperty("Text", "Ala ma kota.");
            config.SetProperty("Bits", new[] { true, false });
            ReadWriteInMemoryStreamTest<YamlConfigurationFormat, TestPropertyBasedConfiguration>(config);
        }

        private bool Has(IEnumerable enumerable, object obj)
        {
            if (enumerable.Cast<object>().Any(el => obj.Equals(el)))
                return true;
            return false;
        }
    }
}

File: SharpOffice.Common.Tests/Configuration/ConfigurationFormatTest.fs
namespace SharpOffice.Common.Tests

open System
open System.Collections.Generic
open System.Linq
open System.IO
open Moq
open NUnit.Framework
open SharpOffice.Common.Configuration
open SharpOffice.Core.Configuration

[<TestFixture>]
[<TestOf(typeof<BinaryConfigurationFormat>)>]
module BinaryConfigurationTest =
    
    let GetMockConfiguration (testData : IEnumerable<KeyValuePair<string, System.Object>>) = 
        let mock = new Mock<IConfiguration>()
        let data = testData.ToList()
        ignore(mock.Setup(fun cfg -> cfg.GetAllProperties()).Returns(data))
        mock.Object

    type TestConfiguration () =
        let dictionary = new Dictionary<string, System.Object>()

        interface IConfiguration with
            member this.GetProperty<'T> (propertyName : string) =
                let property = dictionary.Item(propertyName)
                if not (property :? 'T) then raise (ArgumentException("type T"))
                else property :?> 'T

            member this.SetProperty<'T>( propertyName, propertyValue : 'T) =
                dictionary.Add(propertyName, propertyValue)

            member this.GetAllProperties() = dictionary :> IEnumerable<KeyValuePair<string, Object>>

    [<Test>]
    let WriteAndReadTest () =
        let data = new Dictionary<string, Object>()
        data.Add ("int", 16)
        data.Add ("float", 3.2f)

        use stream = new MemoryStream()
        let configFormat = new BinaryConfigurationFormat()
        configFormat.WriteConfiguration(GetMockConfiguration data, stream)

        stream.Position <- 0L
        let (config : IConfiguration) = configFormat.ReadConfiguration<TestConfiguration>(stream) :> IConfiguration
        for kvp in data do
            Assert.AreEqual(kvp.Value, config.GetProperty(kvp.Key))

[<TestFixture>]
[<TestOf(typeof<YamlConfigurationFormat>)>]
module YamlConfigurationTest =
    
    type TestConfiguration() =
        inherit PropertyBasedConfiguration()
        member val Int = 0 with get,set
        member val Float = 0.0f with get,set

    [<Test>]
    let WriteAndReadTest () =
        let configuration = new TestConfiguration()
        configuration.Int <- 16
        configuration.Float <- 3.2f

        use stream = new MemoryStream()
        let configFormat = new YamlConfigurationFormat()
        configFormat.WriteConfiguration(configuration, stream)

        stream.Position <- 0L
        let config = configFormat.ReadConfiguration<TestConfiguration>(stream)
        Assert.AreEqual(configuration.Int, config.Int)
        Assert.AreEqual(configuration.Float, config.Float)